---
layout: about
title: about
permalink: /
subtitle: 

profile:
  align: right
  image: prof_pic.jpg
  image_circular: false # crops the image to make it circular

selected_papers: false # includes a list of papers marked as "selected={true}"
social: false # includes social icons at the bottom of the page

announcements:
  enabled: true # includes a list of news items
  scrollable: true # adds a vertical scroll bar if there are more than 3 news items
  limit: 3 # leave blank to include all the news in the `_news` folder

latest_posts:
  enabled: true
  scrollable: true # adds a vertical scroll bar if there are more than 3 new posts items
  limit: 3 # leave blank to include all the blog posts
---

<p>
Hi there! I am an <a href="https://www.nserc-crsng.gc.ca/Students-Etudiants/PD-NP/cpra-bprc_eng.asp#a1">NSERC Postdoctoral Fellow</a> at Harvard Medical School and Dana-Farber Cancer Institute. I work with <a href="https://hlilab.github.io/">Heng Li</a>.
</p>

<p>
I obtained my PhD in Math at the University of Toronto, where I was advised by <a href="https://yunwilliamyu.net/content/">Yun William Yu</a>. Previously, I studied Engineering Physics + Mathematics at the University of British Columbia. 
</p>

## Research interests

<!------>

I develop algorithms/tools/theory for [biological sequence analysis](https://en.wikipedia.org/wiki/Sequence_analysis) and its downstream applications. I tend to approach computational biology with a "full-stack approach", starting from [algorithms backed by theory](https://www.genome.org/cgi/doi/10.1101/gr.277637.122) to engineering [widely-used software](https://github.com/bluenote-1577). 

My computational work is driven by my interest in unraveling the latent complexity of microbiomes. The tremendous activity of microbial communities that live around us have important functional consequences. Two common themes define my work:

1. How do we obtain _novel insights_ from microbiome data through _principled_ computational methods?
2. How can we _democratize_ the massive computational analysis of microbial omics data?

For example, see my work on [ultrafast microbial divergence computation](https://www.nature.com/articles/s41592-023-02018-3), [rapid metagenome profiling](https://doi.org/10.1038/s41587-024-02412-y), and [long-read metagenome assembly](https://www.biorxiv.org/content/10.1101/2025.09.05.674543v1). 

