---
layout: about
title: about
permalink: /
subtitle: 

profile:
  align: right
  image: prof_pic.jpg
  image_circular: false # crops the image to make it circular

selected_papers: false # includes a list of papers marked as "selected={true}"
social: false # includes social icons at the bottom of the page

announcements:
  enabled: true # includes a list of news items
  scrollable: true # adds a vertical scroll bar if there are more than 3 news items
  limit: 3 # leave blank to include all the news in the `_news` folder

latest_posts:
  enabled: true
  scrollable: true # adds a vertical scroll bar if there are more than 3 new posts items
  limit: 3 # leave blank to include all the blog posts
---

<p>
Hi there! I am an <a href="https://www.nserc-crsng.gc.ca/Students-Etudiants/PD-NP/cpra-bprc_eng.asp#a1">NSERC Postdoctoral Fellow</a> at Harvard Medical School and Dana-Farber Cancer Institute. I work with <a href="https://hlilab.github.io/">Heng Li</a>.
</p>

<p>
I obtained my PhD in Math at the University of Toronto, where I was advised by <a href="https://yunwilliamyu.net/content/">Yun William Yu</a>. Previously, I studied Engineering Physics + Mathematics at the University of British Columbia. 
</p>

## Research interests

<!------>

I develop algorithms/software/theory for [biological sequence analysis](https://en.wikipedia.org/wiki/Sequence_analysis) and apply these methods to investigate important problems in microbial and microbiome genomics. 

I like to approach computational biology with a "full-stack approach": start from an interesting biology question, create [algorithms backed by theory](https://www.genome.org/cgi/doi/10.1101/gr.277637.122), engineer [widely-used bioinformatics software](https://github.com/bluenote-1577), and then (hopefully!) discover something fundamental about how microbial genomes came to be. 

More specifically, two intertwined themes define my current work:

1. Building sophisticated bioinformatics tools to improve the analysis of massive microbial and meta-omics data.
2. Obtaining new insights about microbiome genome evolution through principled analytical methods.

For example, see my work on [ultrafast microbial divergence computation](https://www.nature.com/articles/s41592-023-02018-3), [rapid metagenome profiling](https://doi.org/10.1038/s41587-024-02412-y), and [long-read metagenome assembly](https://www.biorxiv.org/content/10.1101/2025.09.05.674543v1). 

