---
layout: page
permalink: /publications/
title: publications
description: 
nav: true
nav_order: 2
---

My [Google Scholar profile](https://scholar.google.com/citations?user=hNAwGS8AAAAJ&hl=en) also contains a list of my publications/pre-prints. 

### Selected works

1. **Jim Shaw**, Max Marin, Heng Li. [High-resolution metagenome assembly for diverse long-read technologies with myloasm](https://myloasm-docs.github.io/) (2025). Forthcoming. 
2. **Jim Shaw**, Yun William Yu. [Rapid species-level metagenome profiling and containment estimation with sylph](https://doi.org/10.1038/s41587-024-02412-y) (2024). Nature Biotechnology.
3. **Jim Shaw**, Yun William Yu. [Fast and robust metagenomic sequence comparison through sparse chaining with skani](https://doi.org/10.1038/s41592-023-02018-3) (2023). Nature Methods.
4. **Jim Shaw**, Yun William Yu. [Proving sequence aligners can guarantee accuracy in almost O(m log n) time through an average-case analysis of the seed-chain-extend heuristic](https://genome.cshlp.org/content/early/2023/03/29/gr.277637.122.abstract) (2023). Genome Research.  


### All publications

\* indicates co-lead authors

1. **Jim Shaw**, Heng Li. [High-resolution metagenome assembly for diverse long-read technologies with myloasm](https://myloasm-docs.github.io/) (2025). Forthcoming. 
1. **Jim Shaw**, Christina Boucher, Yun William Yu, Noelle Noyes, Heng Li. [devider: long-read reconstruction of many diverse haplotypes](https://doi.org/10.1101/2024.11.05.621838) (2025). bioRxiv. Accepted to RECOMB2025.
1. Brian Zhang, Grace Oualline, **Jim Shaw**, Yun William Yu. [skandiver: a divergence-based analysis tool for identifying intercellular mobile genetic elements](https://academic.oup.com/bioinformatics/article/40/Supplement_2/ii155/7749064) (2024). ECCB/Bioinformatics. 
1. **Jim Shaw**, Yun William Yu. [Fairy: fast approximate coverage for multi-sample metagenomic binning](https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-024-01861-6) (2024). Microbiome. 
1. **Jim Shaw**\*, Jean-Sebastien Gounot\*, Hanrong Chen, Niranjan Nagarajan, Yun William Yu. [Floria: Fast and accurate strain haplotyping in metagenomes](https://doi.org/10.1093/bioinformatics/btae252) (2024). ISMB/Bioinformatics.
2. **Jim Shaw**, Yun William Yu. [Rapid species-level metagenome profiling and containment estimation with sylph](https://doi.org/10.1038/s41587-024-02412-y) (2024). Nature Biotechnology.
3. **Jim Shaw**, Yun William Yu. [Fast and robust metagenomic sequence comparison through sparse chaining with skani](https://doi.org/10.1038/s41592-023-02018-3) (2023). Nature Methods.
4. Andrew Zheng, **Jim Shaw**, Yun William Yu. [Mora: abundance aware metagenomic read re-assignment for disentangling similar strains](https://doi.org/10.1186/s12859-024-05768-9) (2024). BMC Bioinformatics. 
5. **Jim Shaw**, Yun William Yu. [Proving sequence aligners can guarantee accuracy in almost O(m log n) time through an average-case analysis of the seed-chain-extend heuristic](https://genome.cshlp.org/content/early/2023/03/29/gr.277637.122.abstract) (2023). Genome Research.  
6. Martin Frith, **Jim Shaw**, John Spouge. [How to optimally sample a sequence for rapid analysis](https://doi.org/10.1093/bioinformatics/btad057) (2023). Bioinformatics.  
7. **Jim Shaw**, Yun William Yu. [Theory of local k-mer selection with applications to long-read alignment](https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btab790/6432031) (2022). Bioinformatics.  
8. **Jim Shaw**, Yun William Yu. [Practical probabilistic and graphical formulations of long-read polyploid haplotype phasing](https://doi.org/10.1089/cmb.2021.0436) (2022). Journal of Computational Biology: *RECOMB 2021 Issue*.
9. Tom Ouellette, **Jim Shaw**, Philip Awadalla. [Using image-based haplotype alignments to map global adaptation of SARS-CoV-2](https://www.biorxiv.org/content/10.1101/2021.01.13.426571v1.abstract) (2020). bioRxiv.
10. Ryan Cotsakis\*, **Jim Shaw**\*, Julien Tierny, Joshua Levine. [Implementing Persistence-Based Clustering of Point Clouds in the Topology ToolKit](https://doi.org/10.1007/978-3-030-83500-2_17) (2021). Topological Methods in Data Analysis and Visualization VI. Mathematics and Visualization.
11. Simone Hu\*, Oliver Schnetz\*, **Jim Shaw**\*, Karen Yeats\*. [Further investigations into the graph theory of phi^4-periods and the c2 invariant](https://doi.org/10.4171/aihpd/123) (2020). Annales de l'Institut Henri Poincare D.
12. D. Bertrand, **J. Shaw**, M Narayan, H.Q.A. Ng, S. Kumar, C. Li, M. Dvornicic, J.P. Soldo, J.Y. Kho, O.T. Ng, T. Barkham, B. Young, K. Marimuthu, K.R. Chng, M. Sikic, N. Nagarajan. [Hybrid metagenomic assembly enables high-resolution analysis of resistance determinants and mobile elements in human microbiomes](https://www.nature.com/articles/s41587-019-0191-2) (2019). Nature Biotechnology.

## Talks and presentations

1. [Floria talk](https://github.com/jim-shaw-bluenote/jim-shaw-bluenote.github.io/blob/master/jims_assets/ISMB2024-Floria-JimShaw.pdf). (2024). ISMB 2024, Montreal.
2. [sylph talk v2](https://github.com/jim-shaw-bluenote/jim-shaw-bluenote.github.io/blob/master/jims_assets/GLBIO-TalkJimShaw.pdf). (2024). Great Lakes Bioinformatics 2024, Pittsburgh. [Recorded talk available here](https://www.youtube.com/watch?v=EnZhgq_wzhY&ab_channel=ISCB).
3. [sylph talk v1](https://github.com/jim-shaw-bluenote/jim-shaw-bluenote.github.io/blob/master/jims_assets/sylph_talk.pdf). (2023). Genome Informatics 2023, Cold Spring Harbor Laboratory.
4. [skani talk](https://github.com/jim-shaw-bluenote/jim-shaw-bluenote.github.io/blob/master/jims_assets/2023ISMB_SKANI_TALK.pdf). (2023). ISMB 2023, Lyon.
5. [Seed-chain-extend alignment rigorous average-case analysis talk](https://github.com/jim-shaw-bluenote/jim-shaw-bluenote.github.io/blob/master/jims_assets/recomb2023_talk.pdf). (2023). RECOMB 2023, Istanbul. [Recorded talk available here](https://www.youtube.com/watch?v=zIPIlXy7GM0&ab_channel=RECOMBConferenceSeries) (the slides in the recorded talk are delayed).

