<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Developing sylph - a look into bioinformatics tool development | Jim Shaw </title> <meta name="author" content="Jim Shaw"> <meta name="description" content="Jim Shaw's academic website. "> <meta name="keywords" content="jekyll, jekyll-theme, academic-website, portfolio-website"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://jim-shaw-bluenote.github.io//blog/2024/developing-sylph/"> <script src="/assets/js/theme.js?7ff7ff8a7e8fb7efb59dde1e7b145b6f"></script> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script>initTheme();</script> </head> <body class="fixed-top-nav sticky-bottom-footer"> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">Jim</span> Shaw </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about </a> </li> <li class="nav-item active"> <a class="nav-link" href="/blog/">blog </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications </a> </li> <li class="nav-item "> <a class="nav-link" href="/software/">software </a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">cv </a> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="ti ti-sun-moon" id="light-toggle-system"></i> <i class="ti ti-moon-filled" id="light-toggle-dark"></i> <i class="ti ti-sun-filled" id="light-toggle-light"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="post"> <header class="post-header"> <h1 class="post-title">Developing sylph - a look into bioinformatics tool development</h1> <p class="post-meta"> Created in November 18, 2024 </p> <p class="post-tags"> <a href="/blog/2024"> <i class="fa-solid fa-calendar fa-sm"></i> 2024 </a>   ·   <a href="/blog/category/jekyll"> <i class="fa-solid fa-tag fa-sm"></i> jekyll</a>   <a href="/blog/category/update"> <i class="fa-solid fa-tag fa-sm"></i> update</a> </p> </header> <article class="post-content"> <div id="markdown-content"> <p>In this blog post, I will talk about the process of developing <a href="https://github.com/bluenote-1577/sylph" rel="external nofollow noopener" target="_blank">sylph</a>, a bioinformatics tool we (me and my advisor, <a href="https://yunwilliamyu.net/content/" rel="external nofollow noopener" target="_blank">William</a>) developed. Sylph is now published in <a href="https://www.nature.com/articles/s41587-024-02412-y" rel="external nofollow noopener" target="_blank">Nature Biotechnology</a>.</p> <p>I’ve always found the inner workings of research interesting. This is my contribution. My goal is to demystify tool development in bioinformatics a bit.</p> <h4 id="prelude-what-is-sylph">Prelude: what is sylph?</h4> <p>Sylph is a new method/software that detects what organisms are present in a <a href="https://www.nature.com/articles/nbt.3935" rel="external nofollow noopener" target="_blank">metagenomic DNA sequencing</a> sample of a microbiome. This is sometimes called taxonomic or metagenomic <em>profiling</em>. A metagenomic sequencing produces small DNA fragments of the genomes within a microbiome (e.g. in your gut).</p> <p>This post won’t be about the sylph algorithm, but about its development process. I’ll assume some familiarity with computational metagenomics going forward, but it’s not strictly necessary.</p> <h4 id="part-1---initial-motivation-lets-write-a-faster-mash-late-2022">Part 1 - Initial motivation: let’s write a faster Mash (late 2022)</h4> <hr> <p>The first paper I ever read in the field of bioinformatics was the <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0997-x" rel="external nofollow noopener" target="_blank">Mash</a> paper by Ondov et al. back in 2017 (during an undergraduate internship with <a href="https://www.a-star.edu.sg/gis/our-people/faculty-staff/members/niranjan-nagarajan" rel="external nofollow noopener" target="_blank">Niranjan</a> in Singapore).</p> <p>If you’re unfamiliar with Mash, it’s simply a method for calculating the similarity between two genomes. It does so by analyzing <em>sketches</em>, or small subsets, of k-mers. See <a href="https://genomeinformatics.github.io/mash-screen/" rel="external nofollow noopener" target="_blank">this related blog post</a> for more background. <a href="https://github.com/sourmash-bio/sourmash" rel="external nofollow noopener" target="_blank">Sourmash</a> is a related tool; I recall meeting with <a href="http://ivory.idyll.org/lab/" rel="external nofollow noopener" target="_blank">Titus Brown</a> in 2022 and learned quite a bit about sourmash from him.</p> <p>The Mash algorithm is simple but powerful. It has tremendously influenced how I feel bioinformatics tools should be: <strong>fast, easy to use, and accurate enough for generating biological hypotheses</strong>.</p> <p><strong>Sylph started from a simple curiosity:</strong> when writing <a href="https://www.nature.com/articles/s41592-023-02018-3" rel="external nofollow noopener" target="_blank">a genome-to-genome calculation tool</a> in 2022, I noticed that I could create k-mer sketches faster than Mash. This was done by adapting the k-mer processing routines from <a href="https://github.com/lh3/minimap2" rel="external nofollow noopener" target="_blank">minimap2</a> and learning how to use SIMD instructions to speed stuff up. I thought it would be fun to write a stripped-down, faster k-mer sketching tool for personal use—I didn’t think about publishing at this point. This was built relatively quickly while I was zoning out at a mathematics conference that I attended back home in Vancouver.</p> <h4 id="part-2---stumbling-across-skmer-from-sarmashghi-et-al-mid-late-2022">Part 2 - Stumbling across Skmer from Sarmashghi et al. (mid-late 2022)</h4> <hr> <p>Prior to even starting sylph, I was writing a paper about <a href="https://genome.cshlp.org/content/early/2023/03/29/gr.277637.122" rel="external nofollow noopener" target="_blank">sequence alignment theory</a> in mid-2022. I wanted to justify the usage of a particular statistical model of k-mer statistics.</p> <p>Naturally, I went to <a href="https://doi.org/10.1089/cmb.2021.0431" rel="external nofollow noopener" target="_blank">a familiar paper on k-mer statistics</a> by Blanca et al. to see what they cited, finding the paper <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1632-4" rel="external nofollow noopener" target="_blank">“Skmer: assembly-free and alignment-free sample identification using genome skims”</a> by Sarmashghi et al.</p> <p>I remember skimming through this paper, and while I didn’t read through the math at the time, I understood the basic idea: <strong>low-coverage sequencing creates problems for k-mer sketching</strong>.</p> <p>Back to late-2022, I ran into the following problem while writing my new tool:</p> <h5 id="the-low-abundance-problem-technical-can-skip">The low-abundance problem (technical; can skip)</h5> <hr> <p>Suppose a genome (in some database) <strong>only shares a small fraction of its k-mers within a metagenome sequencing sample</strong>. This could</p> <ul> <li>be due to low sequencing depth in the metagenome (k-mers are not sequenced)</li> </ul> <p><strong>OR</strong></p> <ul> <li>be due to spurious k-mer matches that arise from a different species.</li> </ul> <p>I wanted to compute the <strong>containment average nucleotide identity</strong> (ANI) by using k-mers: this containment ANI generalizes genome-genome ANI to <em>genome-metagenome</em> ANI. If you want to do this using k-mers, sequencing depth becomes an issue.</p> <hr> <p>It turns out that the aforementioned Skmer paper <em>almost</em> tackled the exact problem above, but there was an issue: the Skmer paper only compares genomes-to-genomes, but I needed to compare genomes-to-metagenomes.</p> <p>So during 2022/2023, I spent a few nights at my local McDonald’s, reading through the Skmer paper and working out the rough math for sylph. I implemented the statistical correction method and it seemed to work on some simulated sequences – success!</p> <h4 id="part-3---not-knowing-what-to-do-for-5-months-2022-dec---2023-apr">Part 3 - Not knowing what to do for 5 months (2022 Dec - 2023 Apr)</h4> <hr> <p>The statistical model for sylph was completed in early 2023, but I had no idea what to do with sylph. <strong>I hadn’t figured out how to make it a taxonomic profiler yet.</strong> Let me explain the problem:</p> <h5 id="the-abundance-estimation-problem-technical-can-skip">The abundance estimation problem (technical; can skip)</h5> <hr> <p>If you want the abundance of some organism in your metagenomic sample, most profilers do two steps: (1) classify reads against reference genomes and then (2) count the proportion of reads assigned to each genome.</p> <p>However, if you have (1) two genomes from the same genus, e.g. <em>Escherichia</em>, in your database and (2) a single E. coli read, the read may be similar to multiple <em>Escherichia</em> genomes. How do you “classify” reads in this case? All profilers have some way of dealing with this issue (e.g., Kraken uses a taxonomy).</p> <h2 id="i-didnt-have-a-good-way-of-dealing-with-this-issue-even-for-dereplicated-species-level-databases-because-sylph-doesnt-map-readsit-just-checks-for-k-mers-of-which-many-can-be-shared-between-different-species">I didn’t have a good way of dealing with this issue (even for “dereplicated” species-level databases) because sylph doesn’t map reads—it just checks for k-mers, of which many can be shared between different species.</h2> <p><strong>Back to sylph</strong>: In summary, I had a method that can only check if a genome is contained in a sample (quickly and more accurately than other methods), but <strong>not compute the abundance of the organism</strong> (i.e., do <em>profiling</em>).</p> <p>Funnily enough, I didn’t try hard to tackle this issue. I thought that this was not the “right” problem to solve for sylph. During this time, I had a few other ideas that didn’t pan out. Pulling up an old notebook, I see words such as “k-mer EM algorithm strain resolution” or “optimized sub-linear k-mer indexing”…</p> <h4 id="part-4---try-to-build-around-an-incomplete-algorithm-2023-apr---2023-aug">Part 4 - Try to build around an incomplete algorithm (2023 Apr - 2023 Aug)</h4> <hr> <p>Eventually, in April/May of 2023, I settled on just presenting sylph, but <strong>without the ability to estimate the abundances of organisms</strong>. And by July/August 2023, I wrote a brief communication (~3 pages) based on sylph’s statistical model and showed it could be used for quickly detecting low-coverage genomes.</p> <p>While preparing for manuscript submission, we worked a lot on figuring out how to present the paper. This was rather difficult given that people often only care about abundances.</p> <h4 id="part-5---profiling-winner-take-all-heuristic-saves-the-day-2023-aug">Part 5 - Profiling: winner-take-all heuristic saves the day? (2023 Aug)</h4> <hr> <p>I was planning on submitting sylph in August 2023, and we even had the cover letter ready. However, I was randomly browsing through the sourmash/Mash GitHub one day (I don’t even remember what for), and I came across the “<strong>winner-take-all</strong>” k-mer heuristic. This is explained below:</p> <h5 id="reassigning-k-mers-and-abundance-technical-can-skip">Reassigning k-mers and abundance (technical, can skip)</h5> <hr> <p>The idea behind the winner-take-all heuristic is simple: if a k-mer is shared across many genomes in the database, assign it to the genome with the highest estimated containment ANI. This could solve the abundance problem: if E. coli and E. fergusonii are in your database but only E. coli is present, all of the k-mers for E. fergusonii should be assigned to E. coli. E. fergusonii will have no k-mers left and be thresholded out (see Figure 1 in the sylph paper).</p> <p>To the best of my knowledge, this heuristic is mentioned online (since 2017) but not explicitly explained anywhere, which is why I didn’t run across it. I probably <em>should</em> have thought of this idea. <a href="https://dib-lab.github.io/2020-paper-sourmash-gather/" rel="external nofollow noopener" target="_blank">Irber et al.</a> use a very similar idea that I was familiar with, but the winner-take-all approach made more sense with the statistical model.</p> <hr> <p><strong>Trying the heuristic out:</strong> I didn’t believe this simple heuristic would work, but I decided to give it a try. It took less than an hour to implement, and it worked surprisingly well, so I hit pause on the submission. Comparatively, the statistical model is a more interesting and important contribution, but this was the breakthrough… surprisingly.</p> <p>I finally had the capabilities to estimate organism abundance in August 2023, but I <em>did not</em> want to do the benchmarking necessary for a paper. Benchmarking is much less fun than developing new methods. Unfortunately, the project seemed to have potential.</p> <p>In September 2023, I went to Tokyo to visit <a href="https://sites.google.com/site/mcfrith/" rel="external nofollow noopener" target="_blank">Martin Frith</a> for three months. During these three months, all sylph-related work consisted of benchmarking sylph’s new profiling abilities, creating figures, and rewriting the paper. Recall that it took only 1 hour to implement the winner-take-all heuristic…</p> <p>Luckily, sylph worked well. <em>Why</em> does it work well? Perhaps this is the subject of a talk or another post. Anyways, by late November 2023, I started applying for academic jobs. The benchmarks seemed okay; the paper was tight enough, so we decided to get sylph out as a preprint.</p> <h4 id="part-6---bad-results-on-real-reads-2023-dec">Part 6 - Bad results on real reads? (2023 Dec)</h4> <hr> <p>In December 2023, Florian Plaza Oñate came to me with <a href="https://github.com/bluenote-1577/sylph/issues/5" rel="external nofollow noopener" target="_blank">a bug report</a>, suggesting that sylph was not performing well on real data sets. This was a head-scratching result.</p> <p>As any bioinformatics tool developer knows, your algorithms always perform worse on real data than benchmark data. There will be <em>some</em> aspect of someone’s data that violates your method’s assumptions. However, the data Florian showed me seemed fishy enough that it wasn’t just due to a small biological violation of the model… it seemed systemetically off.</p> <p>A key component of sylph’s model is a <em>stochastic uniformity assumption</em> for read sequencing, leading to Poisson statistics for sequencing. After thinking for a while, I knew this had to be the issue. However, it was not clear if this was an inherent issue with the model, or a technological artefact.</p> <p>An unwritten law in bioinformatics is that if you don’t know what’s happening, you visualize it in the <a href="https://igv.org/doc/desktop/" rel="external nofollow noopener" target="_blank">IGV</a>. So I manually inspected some read alignments, and I found way more duplicated sequences than I expected. It turned out that <strong>PCR duplicates were messing up sylph’s statistical model, violating the Poisson assumption</strong>. Did you know that many Illumina sequencing runs can have &gt; 30% of reads being PCR duplicates? Perhaps embarassingly, I did not.</p> <p>After discovering this issue, I came up with a simple locality-sensitive hashing algorithm for removing PCR duplicates. This seemed to work okay, but I had to rerun a few of my results. This led to an important update and a preprint revision; the revision was also motivated due to people being mad at me for not benchmarking against the latest version of MetaPhlAn :) (lesson learned). This was completed in January 2024.</p> <h4 id="concluding-thoughts-2024-now">Concluding thoughts (2024, now)</h4> <hr> <p>I’m pleasantly surprised at sylph’s reception so far. I’ve had people tell me personally/on social media that they’ve found sylph useful. Regardless of the journal we got into, as a tool developer, I’ve already been very happy with the outcome.</p> <p>Recently, I’m thrilled that sylph has been used for profiling <strong>~2 million sequencing runs</strong> in <a href="https://www.biorxiv.org/content/10.1101/2024.03.08.584059v1" rel="external nofollow noopener" target="_blank">Hunt et al.</a>. Also, apparently Oxford Nanopore Technologies found that <a href="https://a.storyblok.com/f/196663/x/3cff79e4ec/microbeconference_20240502-1333.pdf" rel="external nofollow noopener" target="_blank">sylph was the best method on their long reads</a> as well (see Fig. 3).</p> <h4 id="concluding-opinions-thoughts-and-notes-on-development">Concluding opinions, thoughts, and notes on development</h4> <hr> <p><strong>Development time usage</strong>: Not including paper writing, I estimate I spent 75% of the time on experiments, plotting, and benchmarking compared to 25% developing the method (i.e. writing code, refining the algorithm). I wrote more snakemake, python, and jupyter notebook code than actual sylph code. Why did it take so long?</p> <ol> <li>Figuring out <strong>how</strong> to benchmark is a pain for taxonomic/metagenomic profilers. You have to simulate your own benchmarks, make scripts for normalizing outputs, wrangle other methods’ databases, etc. There are many many potential pitfalls.</li> <li>Figuring out <strong>what</strong> to benchmark is a challenge. This requires a lot of thought.</li> </ol> <p>These time proportions depend on the project. For comparison, sylph’s algorithm is simpler than a genome assembler, which would take more development time. I also think that it’s okay to not spend all your time on software development; I’m a scientist, not a software engineer. A significant part of a tool development project <em>should</em> be figuring out how to benchmark and fit your method into scientific context.</p> <p><strong>Do people care about benchmarks anyway?</strong> Here’s another thing I noticed: many people don’t trust bioinformatics benchmarks. Lots of people fall into two categories:</p> <ul> <li>If a tool’s performance is not critical, the author benchmarks don’t matter that much. People will just play around with your software.</li> <li>If a tool’s performance is critical, you’ll do your own tests instead of trusting author benchmarks; benchmarks from the tool’s authors always place their tool at the top anyways (mysteriously).</li> </ul> <p><strong>Algorithm progress is highly non-linear as a function of time.</strong> It’s hard to predict progress. Furthermore, this project started out way different than how it ended up. I think a more natural approach to method development is correct, rather than shooting for a goal. This means you’re tackling problems in the correct way.</p> <p><strong>Benchmarking taxonomic profilers is hard</strong>. I already <a href="https://jim-shaw-bluenote.github.io/blog/2023/profiling-development/">wrote a post about benchmarking taxonomic profilers</a>. Taxonomic profilers are especially bad because database construction is not intuitive and time consuming.</p> <p><strong>The end</strong>: hopefully this was informative. Feel free to let me know if you have any complaints.</p> </div> </article> <br> <hr> <br> <ul class="list-disc pl-8"></ul> <h2 class="text-3xl font-semibold mb-4 mt-12">Enjoy Reading This Article?</h2> <p class="mb-2">Here are some more articles you might like to read next:</p> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/profiling-development/">Thoughts on metagenomic profiling. Part 1 - as a tool developer</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2022/almost-mlogn-bound-proof-sketch/">A proof sketch of seed-chain-extend runtime being close to O(m log n)</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2022/average-contig-length-n50/">Average contig length times 1.6783469... is equal to N50</a> </li> </div> </div> <footer class="sticky-bottom mt-5" role="contentinfo"> <div class="container"> © Copyright 2024 Jim Shaw. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?e0514a05c5c95ac1a93a8dfd5249b92e"></script> <script defer src="/assets/js/copy_code.js?12775fdf7f95e901d7119054556e495f" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script type="text/javascript">window.MathJax={tex:{tags:"ams"}};</script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/tex-mml-chtml.min.js" integrity="sha256-rjmgmaB99riUNcdlrDtcAiwtLIojSxNyUFdl+Qh+rB4=" crossorigin="anonymous"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script type="text/javascript">function progressBarSetup(){"max"in document.createElement("progress")?(initializeProgressElement(),$(document).on("scroll",function(){progressBar.attr({value:getCurrentScrollPosition()})}),$(window).on("resize",initializeProgressElement)):(resizeProgressBar(),$(document).on("scroll",resizeProgressBar),$(window).on("resize",resizeProgressBar))}function getCurrentScrollPosition(){return $(window).scrollTop()}function initializeProgressElement(){let e=$("#navbar").outerHeight(!0);$("body").css({"padding-top":e}),$("progress-container").css({"padding-top":e}),progressBar.css({top:e}),progressBar.attr({max:getDistanceToScroll(),value:getCurrentScrollPosition()})}function getDistanceToScroll(){return $(document).height()-$(window).height()}function resizeProgressBar(){progressBar.css({width:getWidthPercentage()+"%"})}function getWidthPercentage(){return getCurrentScrollPosition()/getDistanceToScroll()*100}const progressBar=$("#progress");window.onload=function(){setTimeout(progressBarSetup,50)};</script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>addBackToTop();</script> <script src="/assets/js/shortcut-key.js"></script> </body> </html>