<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Thoughts on metagenomic profiling. Part 1 - as a tool developer | Jim Shaw </title> <meta name="author" content="Jim Shaw"> <meta name="description" content="Jim Shaw's academic website. "> <meta name="keywords" content="jekyll, jekyll-theme, academic-website, portfolio-website"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" href="/assets/css/scholar-icons.css?62b2ac103a88034e6882a5be5f3e2772"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://jim-shaw-bluenote.github.io//blog/2023/profiling-development/"> <script src="/assets/js/theme.js?4f125179440d616ec0cbca3e3c2edc42"></script> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script>
    initTheme();
  </script> </head> <body class="fixed-top-nav sticky-bottom-footer"> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">Jim</span> Shaw </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about </a> </li> <li class="nav-item active"> <a class="nav-link" href="/blog/">blog </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications </a> </li> <li class="nav-item "> <a class="nav-link" href="/software/">software </a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">cv </a> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="ti ti-sun-moon" id="light-toggle-system"></i> <i class="ti ti-moon-filled" id="light-toggle-dark"></i> <i class="ti ti-sun-filled" id="light-toggle-light"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="post"> <header class="post-header"> <h1 class="post-title">Thoughts on metagenomic profiling. Part 1 - as a tool developer</h1> <p class="post-meta"> Created on November 24, 2023 </p> <p class="post-tags"> <a href="/blog/2023"> <i class="fa-solid fa-calendar fa-sm"></i> 2023 </a>   ·   <a href="/blog/category/jekyll"> <i class="fa-solid fa-tag fa-sm"></i> jekyll</a>   <a href="/blog/category/update"> <i class="fa-solid fa-tag fa-sm"></i> update</a> </p> </header> <article class="post-content"> <div id="markdown-content"> <h3 id="background---i-spent-a-few-months-developing-a-metagenomic-profiler">Background - I spent a few months developing a metagenomic profiler</h3> <p>I recently spent a few months developing a new metagenomic profiler (shotgun, not 16S) called <a href="https://github.com/bluenote-1577/sylph" rel="external nofollow noopener" target="_blank">sylph</a>. I wanted to write down a few opinions on some difficulties during the development process.</p> <p>The post is a combination of</p> <ul> <li>Opinions for future profiler development choices</li> <li>Pre-rebuttal to reviewers :) (just kidding, somewhat)</li> </ul> <p>I think practitioners may find this post useful for illuminating why profilers are the way they are. Hopefully, you’ll at least learn something about taxonomic profiling.</p> <p><em>Note: I’m open to feedback and opinions, especially if I say something wrong about your tool.</em></p> <h2 id="problem-1---ncbi-resource-usability-for-metagenomic-databases">Problem 1 - NCBI resource usability for metagenomic databases</h2> <p>Here is my understanding of some fundamental aspects of (prokaryotic) databases.</p> <p><strong>RefSeq</strong>: a collection of curated genomes from the NCBI. RefSeq encompasses all kingdoms of life and is comprehensive.</p> <p><strong>NCBI Taxonomy</strong>: an official taxonomy from the NCBI. Provides nomenclature (names of taxa) and their hierarchical relationships.</p> <p><strong>Genome Taxonomy Database (GTDB)</strong>: a database + taxonomy for prokaryotes, distinct from RefSeq/NCBI taxonomy.</p> <p>The default database offered by many profilers, such as <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1891-0" rel="external nofollow noopener" target="_blank">Kraken</a>, uses the NCBI taxonomy + reference genomes from RefSeq. Other methods like <a href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-022-01410-z" rel="external nofollow noopener" target="_blank">mOTUs</a> or <a href="https://www.nature.com/articles/s41587-023-01688-w" rel="external nofollow noopener" target="_blank">MetaPhlAn</a> use the NCBI taxonomy as well, but not just RefSeq genomes. The most popular standardized benchmark, <strong>the <a href="https://www.nature.com/articles/s41592-022-01431-4" rel="external nofollow noopener" target="_blank">CAMI</a> benchmark</strong>, uses RefSeq/NCBI taxonomy.</p> <p>I first want to recognize the amazing work done by the creators of the RefSeq and the NCBI taxonomies. They are invaluable resources. However, I have a few grievances <em>from a purely profiling perspective</em>.</p> <h3 id="lack-of-refseq-database-standardization-across-tools">Lack of RefSeq database standardization across tools</h3> <p>There is no actual “the RefSeq profiling database”. There are options: are genomes <em>complete</em> or <em>scaffold-level</em>? Are prokaryotic genomes <em>representative or reference</em>? Should viral/eukaryotes be included?</p> <p>The lack of standardization is hampered by RefSeq having a continuous release structure as well as a discrete release structure. In many situations, the continuous release is used. For example, the CAMI datasets use a January 8, 2019 RefSeq+Taxonomy snapshot. This is a problem for benchmarking across different versions (discussed later).</p> <h3 id="parsing-ncbi-taxonomy-is-rife-with-pitfalls">Parsing NCBI taxonomy is rife with pitfalls</h3> <p>This is a major grievance I have. The taxonomy that the NCBI provides is <em>difficult to parse</em>. Here are the basic steps:</p> <ol> <li>There is a serialized tree data structure with internal node IDs called “taxid”s called nodes.dmp. You have to parse this tree data structure.</li> <li>You have to get a mapping from a sequence accession to a taxid called accession2taxid. <a href="https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/" rel="external nofollow noopener" target="_blank">But there are many versions</a>. This then involves parsing a &gt; 1GB gzipped line-delimited file.</li> <li>You get the name for the taxid from a file called names.dmp.</li> </ol> <p>The documentation for parsing taxonomy files, getting accession2taxid, etc are stored mostly in READMEs in an FTP server; I found it intimidating. Ultimately, the NCBI taxonomy is meant for all NCBI sequences, <em>not just RefSeq</em>. This makes it comprehensive and an invaluable, complete resource. But you end up mostly parsing non-useful information if you use RefSeq (there are other options too such as NCBI nt).</p> <p><strong>Why should a user care?</strong> Have you ever tried to add new taxonomic nodes to a Kraken database? <a href="https://github.com/DerrickWood/kraken2/issues/436" rel="external nofollow noopener" target="_blank">Not so forgiving</a>. This isn’t the Kraken developers’ fault, it’s because they had to make it concordant with the NCBI taxonomy, which was never really meant to be used for creating profiling databases.</p> <h2 id="problem-2---benchmarking-tools-is-hard-and-no-one-knows-which-tools-are-actually-good">Problem 2 - Benchmarking tools is hard, and no one knows which tools are actually good</h2> <p>A large part of understanding a profiler’s performance comes down to benchmarking. I feel that benchmarking profilers is extremely difficult due to <em>database choice</em>. This will be a constant theme for the following points.</p> <h3 id="poor-synthetic-metagenome-construction-is-easy-and-causes-distrust">Poor synthetic metagenome construction is easy and causes distrust</h3> <p>I’ve seen numerous studies that have constructed synthetic metagenomes and claimed results that don’t seem to match real metagenomes. Some common pitfalls:</p> <ol> <li>uniform abundance distributions</li> <li>genomes in the metagenome are <em>the same</em> as genomes in the chosen databases</li> <li>all genomes in the metagenome have a similar genome in the database.</li> </ol> <p>Some benchmark studies claim near-perfect accuracy for many tools, whereas other benchmarks claim most tools perform poorly. This is due to many factors: parameters, database choice, construction of metagenome, etc.</p> <p>For me, this has created distrust of custom synthetic benchmarks. I find myself glazing over custom benchmarks and just looking at CAMI results… but this is not a good thing for the field. While standardized datasets are great, they must lack <em>some</em> aspect of a real metagenome. Also, methods will inevitably overfit to them.</p> <h3 id="how-do-you-choose-a-database-when-comparing-methods">How do you choose a database when comparing methods?</h3> <p>mOTUs and MetaPhlAn use the NCBI taxonomy, but they use a fixed version since their database is not customizable. mOTUs3 <a href="https://github.com/motu-tool/mOTUs/issues/85" rel="external nofollow noopener" target="_blank">seems to use the 8 January 2019</a> for concordance with CAMI, probably, and I believe MetaPhlAn4 <a href="https://forum.biobakery.org/t/which-ncbi-taxdump-version-used-for-metaphlan4-database/4989" rel="external nofollow noopener" target="_blank">depends on the exact version used</a>, but not the 8 Jan 2019 version.</p> <p>When you are comparing tools, there are three options.</p> <ol> <li> <strong>(A)</strong> You choose a default database for each tool and compare across databases.</li> <li> <strong>(B)</strong> You use <em>the same</em> database+taxonomy for every method.</li> <li> <strong>(C)</strong> You use the the same <em>taxonomy</em> but different databases.</li> </ol> <p>Option <strong>(B)</strong> seems mostly fair and is done by many manuscripts. But some tools were designed with certain database structures in mind. Kraken2, for example, prefers a redundant database over a dereplicated one.</p> <p>Option <strong>(A), (C)</strong> possibly represents a more accurate representation of actual usage – I bet folks out there use Kraken2’s RefSeq default database and MetaPhlAn4’s default marker database, which are completely different.</p> <p>But <strong>(A)</strong> is an issue due to the following:</p> <ul> <li>One method could use a more comprehensive database or simply choose genomes that are more similar to the metagenome. Which profiler is really better?</li> <li> <strong>Taxa names can change between different databases/taxonomy versions.</strong> Comparisons become concerning across taxonomy versions.</li> </ul> <p>Option <strong>(C)</strong> suffers from the first problem above, but not the second.</p> <h3 id="benchmarking-becomes-a-headache">Benchmarking becomes a headache</h3> <p>What’s a good solution? To be honest, I don’t know. Some thoughts:</p> <ul> <li>Marker gene methods are extremely performant. But benchmarking against them means using their fixed taxonomies and databases.</li> <li>Many people are using GTDB now (see <a href="https://www.biorxiv.org/content/10.1101/712166v1" rel="external nofollow noopener" target="_blank">here</a>, for example). I won’t get into NCBI vs GTDB here.</li> <li>I tried to map MetaPhlAn4 taxonomic profiles to the GTDB, but NCBI and GTDB are inherently not 1-to-1, so issues arose.</li> <li>Even if you use the “NCBI taxonomy”, are you comparing across versions?</li> </ul> <p>I am now cautious when I interpret results between MetaPhlAn/mOTUs and non-marker gene methods. Many of the studies that I’ve read seem to lack information about how cross-taxonomy results were obtained. The study <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2021.643682/full" rel="external nofollow noopener" target="_blank">here</a> by Parks et al. is an exception and did a really great job of outlining its benchmarking procedure.</p> <h2 id="problem-3----standardizing-taxonomic-profiling-outputs">Problem 3 - Standardizing taxonomic profiling outputs</h2> <p>Software output their profiling results in different formats, and it’s up to the user to standardize them. This is annoying for users. I’ll discuss why I think this situation ended up happening.</p> <h3 id="methods-do-different-things">Methods do different things</h3> <ol> <li> <p><strong>Taxonomic profiling</strong> is the quantification of taxa for a metagenomic sample. Taxonomic profiling, of course, requires a taxonomy.</p> </li> <li> <p>I will define <strong>genome profiling</strong> (or metagenome profiling) as the quantification of the genomes in a database. Note that genome profiling does not require a taxonomy, but genome profiles can be turned into a taxonomic profile by incorporating a taxonomy.</p> </li> <li> <p>Some methods are <strong>sequence classifiers</strong>. They classify each sequence (i.e. read) against a database and then use this for either taxonomic or genome quantification. The term <strong>taxonomic binner</strong> is used; I dislike the term because it gets confused with contig binning.</p> </li> </ol> <p>Sequence classifiers, genome profilers, and taxonomic profilers are not exclusive; most methods do multiple things. As such, they will output different things.</p> <p>For example, Kraken does (1) and (3). Sylph does (2) and can incorporate taxonomic information downstream for (1). MetaPhlAn4 does (1), but algorithmically it’s more like (2) – it <a href="https://forum.biobakery.org/t/origin-clade-specific-marker-genes/3806" rel="external nofollow noopener" target="_blank">aligns reads to a species-level database of marker genes</a> (the “genomes”) and then sums up abundances.</p> <h3 id="we-need-a-standardized-taxonomic-profiling-output">We need a standardized taxonomic profiling output</h3> <p>Standardizing sequence classifiers and genome profiling is out of scope, but we can at least attempt to standardize <strong>taxonomic profiling</strong> outputs. The <a href="https://github.com/CAMI-challenge/contest_information/blob/master/file_formats/CAMI_TP_specification.mkd" rel="external nofollow noopener" target="_blank">CAMI format for taxonomic profiling</a> is a possible choice, but it is not the standard for Kraken or MetaPhlAn. MetaPhlAn has its own output, which is similar.</p> <p>I think these formats are reasonable, but here are some rudimentary thoughts.</p> <ul> <li>I don’t like taxids included in a format specification. This requires a relationship to the NCBI taxonomy. They could be included as additional tags.</li> <li>Existing output formats do not allow for customization. The <a href="https://samtools.github.io/hts-specs/SAMv1.pdf" rel="external nofollow noopener" target="_blank">SAM format</a> allows for optional tags for each record to record additional information, for example.</li> <li>We should really be attaching some sort of confidence score to each classification. Think about how useful MAPQs are in read mapping.</li> <li>We should record the exact taxonomies (i.e. versions) used. Taxonomic profiles and taxonomy are intertwined.</li> </ul> <h2 id="conclusion">Conclusion</h2> <p>Building, using, and understanding taxonomic profilers is surprisingly difficult. The differences between methods on (1) database, (2) taxonomy, and (3) outputs makes interpreting benchmarks impossible. Since developing a profiler, I have a newfound respect for the people behind large standardized benchmarks, especially the folks behind CAMI for tackling these challenges.</p> <p>My ideal future as a tool developer / user?</p> <ol> <li>Let’s have a collection of standardized and versioned databases, each with (1) a set of genomes and (2) an associated taxonomy file that is easy to parse.</li> <li>We can extend each database by simply dropping in a fasta file and adding a new line in the taxonomy file.</li> <li>Let’s allow each tool to use one of these standardized databases and output standardized taxonomic profiles, making benchmarking and cross-profile comparison easy. This allows practitioners to try different databases easily too.</li> </ol> <h3 id="example-taxonomy-format-solution">Example taxonomy format solution</h3> <p>Here is a simple format: store a correspondence between each genome file/sequence and a string in a two-column TSV file:</p> <p>Genome_file Tax_string</p> <p>GCA_945889495.1.fasta d__Bacteria;p__Desulfobacterota_B;…;g__DP-20;s__DP-20 sp945889495</p> <p>GCA_934500585.1.fasta d__Bacteria;p__Bacillota_A;…;g__RQCD01;s__RQCD01 sp008668455</p> <p>That’s it – just one file. This would allow you to reconstruct a taxonomic tree, giving you <em>the exact same information</em> as the NCBI files.</p> <p>This would allow your favourite MAG classification tool (GTDB-TK, for example) to be integrated into a taxonomic profiler database. To add a new species, you just add a genome and its new taxonomy string to the file. This is roughly, but not approximately, <a href="https://github.com/bluenote-1577/sylph/wiki/Integrating-taxonomic-information-with-sylph" rel="external nofollow noopener" target="_blank">what sylph does</a>.</p> <p><strong>Acknowledgements:</strong> Thanks to the people in my <a href="https://bsky.app/profile/jimshaw.bsky.social/post/3ka5ftbdgbt2f" rel="external nofollow noopener" target="_blank">bluesky complaint thread</a> who gave their thoughts on profilers.</p> <p><strong>Complaints/Thoughts/Inaccuracies?:</strong> Feel free to e-mail me or DM me at my twitter @jim_elevator.</p> </div> </article> <br> <hr> <br> <ul class="list-disc pl-8"></ul> <h2 class="text-3xl font-semibold mb-4 mt-12">Enjoy Reading This Article?</h2> <p class="mb-2">Here are some more articles you might like to read next:</p> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2024/developing-sylph/">Developing sylph - a look into bioinformatics tool development</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2022/almost-mlogn-bound-proof-sketch/">A proof sketch of seed-chain-extend runtime being close to O(m log n)</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2022/average-contig-length-n50/">Average contig length times 1.6783469... is equal to N50</a> </li> </div> </div> <footer class="sticky-bottom mt-5" role="contentinfo"> <div class="container"> © Copyright 2026 Jim Shaw. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js?a0db7e5d5c70cc3252b3138b0c91dcaf" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?e0514a05c5c95ac1a93a8dfd5249b92e"></script> <script defer src="/assets/js/copy_code.js?c8a01c11a92744d44b093fc3bda915df" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script> <script src="/assets/js/mathjax-setup.js?a5bb4e6a542c546dd929b24b8b236dfd"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script defer src="/assets/js/progress-bar.js?2f30e0e6801ea8f5036fa66e1ab0a71a" type="text/javascript"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>
    addBackToTop();
  </script> </body> </html>